src/hugr/validate.rs

```diff
@@ -558,6 +521,82 @@ impl<'a, 'b> ValidationContext<'a, 'b> {
         }
         .into())
     }
+
+    fn validate_subtree(
+        &mut self,
+        node: Node,
+        var_decls: &[TypeParam],
+    ) -> Result<(), ValidationError> {
+        let op_type = self.hugr.get_optype(node);
+        // The op_type must be defined only in terms of type variables defined outside the node
+        // TODO consider turning this match into a trait method?
+        match op_type {
+            OpType::LeafOp(crate::ops::LeafOp::CustomOp(b)) => {
+                // Try to resolve serialized names to actual OpDefs in Extensions.
+                let temp: ExternalOp;
+                let resolved = match &**b {
+                    ExternalOp::Opaque(op) => {
+                        // If resolve_extension_ops has been called first, this would always return Ok(None)
+                        match resolve_opaque_op(node, op, self.extension_registry)? {
+                            Some(exten) => {
+                                temp = ExternalOp::Extension(exten);
+                                &temp
+                            }
+                            None => &**b,
+                        }
+                    }
+                    ExternalOp::Extension(_) => &**b,
+                };
+                // Check TypeArgs are valid, and if we can, fit the declared TypeParams
+                match resolved {
+                    ExternalOp::Extension(exten) => exten
+                        .def()
+                        .validate_args(exten.args(), self.extension_registry, var_decls)
+                        .map_err(|cause| ValidationError::SignatureError { node, cause })?,
+                    ExternalOp::Opaque(opaq) => {
+                        // Best effort. Just check TypeArgs are valid in themselves, allowing any of them
+                        // to contain type vars (we don't know how many are binary params, so accept if in doubt)
+                        for arg in opaq.args() {
+                            arg.validate(self.extension_registry, var_decls)
+                                .map_err(|cause| ValidationError::SignatureError { node, cause })?;
+                        }
+                    }
+                }
+            }
+            OpType::LeafOp(crate::ops::LeafOp::TypeApply { ta }) => {
+                ta.validate(self.extension_registry)
+                    .map_err(|cause| ValidationError::SignatureError { node, cause })?;
+            }
+            _ => (),
+        }
+
+        // Check port connections.
+        for dir in Direction::BOTH {
+            for (i, port_index) in self.hugr.graph.ports(node.pg_index(), dir).enumerate() {
+                let port = Port::new(dir, i);
+                self.validate_port(node, port, port_index, op_type, var_decls)?;
```

Comment:
Say we have a ModuleOp "outer" containing an op "x" and a nested module "inner" which contains a node "y". Each module is free to bind some type variables. Can we have an inter graph edge between "x" and "y"? If so, the current de Bruijn indices aren't sufficient.

If this is true and you agree, I'll make an issue for it

Reply:
Ouch! You evil.... ;-) ;-)

I think though that you couldn't have an inter-graph/non-local Value edge. The only thing you could have would be a Static edge. But yeah, if such can refer to enclosing type variables, that would lead to exactly the kind of problem you describe (?)

This is sounding like the death knell of "nested Function nodes can refer to type variables declared by an enclosing FuncDefn"....

Reply:
Good spot!!

Reply:
Ok, I've added extra restriction on Static Edges, to rule out the case of polymorphic consts which would otherwise break this. (Also ruled out non-closed FuncDefn's by replacing rather than concatenating lists of binders, which dealt with the function case, but needed to go further for Const.)

Reply:
I was going to suggest instantiating variables with a different naming scheme during validation rather than scrapping the concept, but this might be a more appropriate level of abstraction for an IR (and we can always change our minds later)